File: Trab2/ClientGenerator.cpp
#include <iostream>
#include <unistd.h>
#include <cstdlib>
#include "ClientGenerator.h"

// constructor
ClientGenerator::ClientGenerator(BarberShop *bs, unsigned int max): shop(bs), maxClientsTG(max) {}

// destructor
ClientGenerator::~ClientGenerator() {
    shop = nullptr;
}

// client generating limit
unsigned int ClientGenerator::getMaxClientsTG() {
    return maxClientsTG;
}

// the thread start method
void ClientGenerator::start() {
    pthread_attr_t attr;
    pthread_attr_init(&attr);
    pthread_create(&tid, &attr, run, this);
    return;
}

// joint the ClientGenerator
void ClientGenerator::joinClientGeneratorThread() {
    pthread_join(tid, nullptr);
}

// overriding Runnable run() method
void* ClientGenerator::run(void *arg) {

    ClientGenerator *cg = (ClientGenerator*) arg;
    if (nullptr == cg) {
        std::cout << "Error! Invalid ClientGenerator pointer";
        pthread_exit(nullptr);
    }

    std::cout << std::endl << "---- Haired clients comming!! ----" << std::endl;

    unsigned int i = 0;

    Client *c = nullptr;
    // see maxClients
    while(i < cg->getMaxClientsTG()) {

        // create a Client Thread
        c = new Client(cg->shop, i+1);

        c->start();

        // try to sleep 700 ms
        usleep(rand()%400000 + 1);
        // see maxClients
        i++;
    }

    c = nullptr;
    std::cout << "Done! We are closing now! But all remaining clients are going to be attended" << std::endl;
    cg->shop->close();
    pthread_exit(nullptr);

}

File: Trab2/ClientGenerator.java
import java.util.*;
// internal class, a Client object factory
public class ClientGenerator implements Runnable {

    // client threads list
    LinkedList<Thread> ts;

    // the shop
    BarberShop shop;

    // number of customers
    int maxClients;

    // constructor
    public ClientGenerator(BarberShop bs, int max) {
        shop = bs;
        maxClients = max;
        ts = new LinkedList<Thread>();
    }

    // overriding Runnable run() method
    @Override
    public void run() {
        System.out.println("\nHaired clients comming!!");

        // randon integer generator
        Random rand = new Random();

        int i = 0;

        // see maxClients
        while(i < maxClients*2) {

            // create a Client Thread
            Client c = new Client(shop, i+1);
            Thread t = new Thread(c);

            // add Thread to BarberShop clients list
            ts.add(t);
            // run the Client Thread
            t.start();

            // try to sleep 400 ms
            try {
                Thread.sleep(rand.nextInt(400));
            } catch (InterruptedException ieerror){
                System.out.println("Error: " + ieerror);
            }
            // see maxClients
            i++;
        }

        // join all client threads
        for (i = 0; i < ts.size(); i++) {
            Thread t = ts.poll();
            if (null != t) {
                try {
                    // wait for Client Thread
                    t.join();
                } catch (InterruptedException ieerror) {
                    System.out.println("Error: " + ieerror);
                }
            }
        }
    }
}
